java-extras/src/main/java/org/triplea/java/cache/ExpiringAfterWriteCache.java

```diff
@@ -0,0 +1,90 @@
+package org.triplea.java.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import org.triplea.java.timer.ScheduledTimer;
+import org.triplea.java.timer.Timers;
+
+/**
+ * Cache that expires values when a TTL (time to live) expires. TTL timer starts when the value is
+ * written and is renewed if the value is 'refreshed'. The cache will reliably invoke a
+ * 'removeListener' at least once when cache items are removed or expired.
+ *
+ * @param <IdT> Type that identifies the keys of the map.
+ * @param <ValueT> Type that is placed as a value in the map.
+ */
+public class ExpiringAfterWriteCache<IdT, ValueT> implements TtlCache<IdT, ValueT> {
+
+  private final Cache<IdT, ValueT> cache;
+  private final ScheduledTimer cleanupTimer;
+  private final Consumer<CacheEntry<IdT, ValueT>> removalListener;
+
+  @SuppressWarnings("unchecked")
+  public ExpiringAfterWriteCache(
+      final long duration,
+      final TimeUnit timeUnit,
+      final Consumer<CacheEntry<IdT, ValueT>> removalListener) {
+    cache =
+        Caffeine.newBuilder()
+            .expireAfterWrite(duration, timeUnit)
+            .removalListener(
+                (key, value, cause) -> {
+                  if (cause == RemovalCause.EXPIRED || cause == RemovalCause.EXPLICIT) {
+                    removalListener.accept(new CacheEntry<>((IdT) key, (ValueT) value));
```

Comment:
What about using BiConsumer instead?

Reply:
Just had a look at the caffeine documentation because I thought the explicit casting was weird.
You can avoid (I'm 99% sure) the unchecked warning by using
```java
.<IdT, ValueT>removalListener((key, value, cause) -> {});
```
Instead

Reply:
You could declare the types on the lambda parameters instead,
```java
.removalListener((IdT key, ValueT value, RemovalCause cause) -> {});
```